File: common/src/main/java/com/thewizrd/common/weatherdata/WeatherResult.java
package com.thewizrd.common.weatherdata;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.thewizrd.shared_resources.weatherdata.model.Weather;

public final class WeatherResult {
    private boolean isSavedData = false;
    private Weather weather = null;

    public boolean isSavedData() {
        return isSavedData;
    }

    @Nullable
    public Weather getWeather() {
        return weather;
    }

    private WeatherResult() {
    }

    @NonNull
    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    static WeatherResult create(@Nullable Weather weather, boolean freshFromProvider) {
        WeatherResult result = new WeatherResult();
        result.isSavedData = !freshFromProvider;
        result.weather = weather;
        return result;
    }
}

File: common/src/main/java/com/thewizrd/common/weatherdata/WeatherResult.kt
package com.thewizrd.common.weatherdata

import com.thewizrd.shared_resources.exceptions.WeatherException
import com.thewizrd.shared_resources.weatherdata.model.Weather

sealed class WeatherResult {
    open val data: Weather? = null

    data class Success(override val data: Weather, val isSavedData: Boolean = false) :
        WeatherResult()

    data class WeatherWithError(
        override val data: Weather,
        val isSavedData: Boolean = true,
        val exception: WeatherException
    ) : WeatherResult()

    data class NoWeather(override val data: Weather? = null, val isSavedData: Boolean = false) :
        WeatherResult()

    data class Error(val exception: WeatherException) : WeatherResult()
}

fun Weather?.toWeatherResult(isSavedData: Boolean = false): WeatherResult {
    return if (this == null) {
        WeatherResult.NoWeather(isSavedData = isSavedData)
    } else {
        WeatherResult.Success(this, isSavedData)
    }
}